# Табличные данные

**Цель**: провести анализ табличных данных с помощью средств языка программирования Python

**Выполнил:** ст.группы 24ВВИ1 Будников А.С.

**TODO:** написать ответы на теорию, сказать о признаке Soil_Type и др "бинарных" в анализе

## Теоретическая часть

### 1. Что называют табличными данными?

Табличные данные — это структурированный набор информации, или данные, которые можно представить в виде таблицы, где строки — это отдельные записи, а столбцы — различные характеристики исследуемого объекта.

### 2. Чем табличные данные отличаются от: изображений; текста; временных рядов?

  * В изображениях важна пространственная структура: пиксели образуют сетку, и соседние точки связаны между собой, из чего складывается картинка. Поэтому для обработки нужны модели, которые учитывают это соседство.
  * Текст представляет собой последовательность слов переменной длины, в которой смысл во многом зависит от порядка и контекста, в отличие от таблиц, где порядок, в целом, не важен.
  * Временные ряды это те же табличные данные, но упорядоченные по времени, где каждое значение зависит от предыдущих. Порядок здесь критичен для прогноза и поиска аномалий, а если значения перемешать — информация потеряется.


### 3. Почему большинство промышленных ML-задач работают именно с табличными данными?

На это есть ряд причин, в центре которых стоит бизнес-логика. Чаще всего компании уже хранят необходимые данные в виде таблиц, то есть реляционных баз данных, а признаки в таких таблицах часто осмыслены и структурированы. Также по таблице легче объяснить поведение модели и встроить в уже существующие системы.

### 4. Какие типы признаков встречаются в табличных данных?

Выделяются 4 основных типа - числовые, категориальные, временные и текстовые. Категориальные и текстовые кодируются по одному из существующих методов, то есть переводятся в числовой формат, а временные можно разбить на несколько признаков, например, день недели, месяц и т.п.

### 5. Что такое признак и целевая переменная?

* Признак - это характеристика исследуемого объекта, используемая как входные данные для совершения моделью прогноза, или индивидуальное измеримое свойство или характеристика наблюдаемого явления. Выбор информативных, отличительных и независимых признаков напрямую влияет на эффективность алгоритма.
* Целевая переменная - индивидуальное измеримое свойство или характеристика наблюдаемого явления или объекта, которое модель машинного обучения пытается предсказать на основе входных данных (признаков), от которых она зависит.

### 6. Различия между: числовыми; категориальными; бинарными признаками.

Числовые признаки принимают значения из непрерывного или дискретного диапазона, в то время как категориальные описывают принадлежность объекта к какой-либо выделяемой группе, часто кодируются в виде целых чисел. Над категориальными проводить арифметические операции не имеет смысла, так как видимого результат данная процедура не имеет. Бинарные признаки в какой-то степени можно назвать частным случаем категориальных данных; они имеют два потенциальных значения - истина/ложь или 0/1 в зависимости от интерпретации; не требуют кодирования, как и числовые.

### 7. Почему корректное кодирование признаков критично для моделей ML?

Преобразования данных — важный шаг в подготовке ваших данных к задачам машинного обучения, так как:
* Алгоритмы машинного обучения ожидают данные в определённом формате, обычно числовом.
* Правильное кодирование данных гарантирует, что алгоритмы могут извлекать значимые закономерности и связи из данных.
* Отсутсвие приведения числовых признаков с большим разбросом к единому масштабу может привести к доминированию признаков с большей величиной.
* Неправильное применение алгоритмов кодирования может привести к значительному росту времени обучения и переобучения на редких категориях.
* Несоглсованность разных типов выборок может привести к неправильной работе модели.

In [1]:
import pandas as pd
import numpy as np

## Анализ

На данном шаге были импортированы необходимые для работы библиотеки:

  - pandas - библиотека для работы с табличными данными
  - numpy - библиотека для работы с массивами

In [2]:
df = pd.read_csv('data/covtype.csv') # default delimiter is ','
df.head() # default is 5

,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Hillshade_9am,Hillshade_Noon,Hillshade_3pm,Horizontal_Distance_To_Fire_Points,...,Soil_Type32,Soil_Type33,Soil_Type34,Soil_Type35,Soil_Type36,Soil_Type37,Soil_Type38,Soil_Type39,Soil_Type40,Cover_Type
0,2596,51,3,258,0,510,221,232,148,6279,...,0,0,0,0,0,0,0,0,0,5
1,2590,56,2,212,-6,390,220,235,151,6225,...,0,0,0,0,0,0,0,0,0,5
2,2804,139,9,268,65,3180,234,238,135,6121,...,0,0,0,0,0,0,0,0,0,2
3,2785,155,18,242,118,3090,238,238,122,6211,...,0,0,0,0,0,0,0,0,0,2
4,2595,45,2,153,-1,391,220,234,150,6172,...,0,0,0,0,0,0,0,0,0,5


In [3]:
print(df.shape)

(581012, 55)


## Анализ

В качестве датасета используется [Forest Cover Types](https://www.kaggle.com/datasets/uciml/forest-cover-type-dataset/data).
Здесь содержатся данные о типах лесного покрова для территорий национального парка Рузвельта в Колорадо США. Каждая строка таблицы описывает один участок земли площадью 30×30 метров. Набор признаков включает информацию о типе дерева, степени затенения, расстоянии до близлежащих ориентиров (дорог и т. д.), типе почвы и местном рельефе. 

Всего в датасете 581 012 записей и 55 столбцов как показано выше.

In [4]:
print(df.dtypes)

Elevation                             int64
Aspect                                int64
Slope                                 int64
Horizontal_Distance_To_Hydrology      int64
Vertical_Distance_To_Hydrology        int64
Horizontal_Distance_To_Roadways       int64
Hillshade_9am                         int64
Hillshade_Noon                        int64
Hillshade_3pm                         int64
Horizontal_Distance_To_Fire_Points    int64
Wilderness_Area1                      int64
Wilderness_Area2                      int64
Wilderness_Area3                      int64
Wilderness_Area4                      int64
Soil_Type1                            int64
Soil_Type2                            int64
Soil_Type3                            int64
Soil_Type4                            int64
Soil_Type5                            int64
Soil_Type6                            int64
Soil_Type7                            int64
Soil_Type8                            int64
Soil_Type9                      

In [5]:
print(df.nunique())
categorical_cols = [col for col in df.columns if df[col].nunique() == 2]
numeric_cols = [col for col in df.columns if df[col].nunique() > 2 and col != 'Cover_Type']

Elevation                             1978
Aspect                                 361
Slope                                   67
Horizontal_Distance_To_Hydrology       551
Vertical_Distance_To_Hydrology         700
Horizontal_Distance_To_Roadways       5785
Hillshade_9am                          207
Hillshade_Noon                         185
Hillshade_3pm                          255
Horizontal_Distance_To_Fire_Points    5827
Wilderness_Area1                         2
Wilderness_Area2                         2
Wilderness_Area3                         2
Wilderness_Area4                         2
Soil_Type1                               2
Soil_Type2                               2
Soil_Type3                               2
Soil_Type4                               2
Soil_Type5                               2
Soil_Type6                               2
Soil_Type7                               2
Soil_Type8                               2
Soil_Type9                               2
Soil_Type10

## Анализ

Как видно из двух верхних пунктов таблица содержит данные типа ``int64``, поэтому выделить категориальные признаки здесь представляется возможным только путём отсечения столбцов по количеству уникальных значений. В данном случае пороговым числом было выбрано **2**. В данном датасете эвристика работает хорошо, потому что все категориальные признаки являются бинарными, и разрыв между группами очень велик.

Такой подход во многом оправдан: признаки с малым числом уникальных значений с высокой вероятностью являются бинарными или порядковыми категориями, тогда как признаки с большим числом уникальных значений - это непрерывные числовые величины. Важно также исключить из списка целевую переменную *Cover_Type*, имеющую 7 уникальных значений.

В результате разделения были получены две группы:

  - Числовые признаки (10 столбцов): Elevation, Aspect, Slope, Horizontal_Distance_To_Hydrology, Vertical_Distance_To_Hydrology, Horizontal_Distance_To_Roadways, Hillshade_9am, Hillshade_Noon, Hillshade_3pm, Horizontal_Distance_To_Fire_Points
  - Категоральные признаки (45 столбцов): Wilderness_Area1 … Wilderness_Area4 (бинарные), Soil_Type1 … Soil_Type40 (бинарные)

Категоральные признаки в данном случае представлены как результат One-Hot Encoding, поэтому вместо 40 значений для единого Soil_Type, имеем 40 уникальных столбцов, которые по сути являются одним категориальным признаком, но на уровне представления в таблице будем воспринимать их как 40 бинарных столбцов.

In [6]:
df[numeric_cols].describe().T[["max", "min", "mean"]]

,max,min,mean
Elevation,3858.0,1859.0,2959.365301
Aspect,360.0,0.0,155.656807
Slope,66.0,0.0,14.103704
Horizontal_Distance_To_Hydrology,1397.0,0.0,269.428217
Vertical_Distance_To_Hydrology,601.0,-173.0,46.418855
Horizontal_Distance_To_Roadways,7117.0,0.0,2350.146611
Hillshade_9am,254.0,0.0,212.146049
Hillshade_Noon,254.0,0.0,223.318716
Hillshade_3pm,254.0,0.0,142.528263
Horizontal_Distance_To_Fire_Points,7173.0,0.0,1980.291226


In [7]:
# признаки с наибольшим разбросом.
var_vals = df[numeric_cols].std().sort_values(ascending=False)
var_vals

Horizontal_Distance_To_Roadways       1559.254870
Horizontal_Distance_To_Fire_Points    1324.195210
Elevation                              279.984734
Horizontal_Distance_To_Hydrology       212.549356
Aspect                                 111.913721
Vertical_Distance_To_Hydrology          58.295232
Hillshade_3pm                           38.274529
Hillshade_9am                           26.769889
Hillshade_Noon                          19.768697
Slope                                    7.488242
dtype: float64

## Анализ

После разделения признаков на числовые и категориальные была рассчитана статистика для числовых признаков: минимум, максимум, среднее и разброс.

Были сделаны следующие наблюдения: 

  - Vertical_Distance_To_Hydrology — единственный признак с отрицательными значениями
  - Наибольший размах — у Horizontal_Distance_To_Fire_Points (7173) и Horizontal_Distance_To_Roadways (7117)
  - Наиболее стабильные признаки: Slope (7.49), Hillshade_Noon (19.77), Hillshade_9am (26.77)
  - В то время как Horizontal_Distance_To_Roadways (1559.25), Horizontal_Distance_To_Fire_Points (1324.20), Elevation (279.98) имеют наибольший разброс

Расстояния до инфраструктуры и высота сильно варьируются от участка к участку, тогда как крутизна (Slope) и освещённость (Shade) более однородны. Признаки имеют разный масштаб, что важно учитывать.

Выполните фильтрацию данных по одному условию (например, значение признака больше/меньше заданного).

In [8]:
CONDITION_NUMBER = 250
df_filtered = df[df['Hillshade_3pm'] > CONDITION_NUMBER]
df_filtered

,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Hillshade_9am,Hillshade_Noon,Hillshade_3pm,Horizontal_Distance_To_Fire_Points,...,Soil_Type32,Soil_Type33,Soil_Type34,Soil_Type35,Soil_Type36,Soil_Type37,Soil_Type38,Soil_Type39,Soil_Type40,Cover_Type
229337,2719,271,42,509,236,912,80,213,251,1741,...,0,0,0,0,0,0,0,0,0,2
325592,3355,274,44,162,65,3360,72,207,251,1727,...,0,0,0,0,0,0,0,1,0,1
326286,3355,268,43,134,49,3372,79,214,251,1697,...,0,0,0,0,0,0,0,1,0,1
352621,2836,268,43,300,202,1358,81,215,251,1482,...,0,0,0,0,0,0,0,0,0,2
353237,2804,264,44,270,170,1358,79,215,251,1443,...,0,0,0,0,0,0,0,0,0,2
353238,2833,266,44,300,199,1379,76,213,252,1471,...,0,0,0,0,0,0,0,0,0,2
353242,2913,265,42,420,279,1465,83,217,251,1584,...,0,1,0,0,0,0,0,0,0,2
353855,2801,269,45,270,167,1380,71,209,252,1433,...,0,0,0,0,0,0,0,0,0,2
353856,2833,271,45,300,199,1401,70,208,252,1461,...,0,0,0,0,0,0,0,0,0,2
353860,2907,268,44,420,273,1486,73,211,252,1575,...,0,1,0,0,0,0,0,0,0,2


## Анализ

Сортировка была выполнена по условию: числовой признак Hillshade_3pm больше 250. 

В результате получаем 39 записей с наивысшей освещенностью

Отсортируйте отфильтрованные данные: по возрастанию

In [9]:
df_filtered.sort_values('Hillshade_3pm')

,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Hillshade_9am,Hillshade_Noon,Hillshade_3pm,Horizontal_Distance_To_Fire_Points,...,Soil_Type32,Soil_Type33,Soil_Type34,Soil_Type35,Soil_Type36,Soil_Type37,Soil_Type38,Soil_Type39,Soil_Type40,Cover_Type
229337,2719,271,42,509,236,912,80,213,251,1741,...,0,0,0,0,0,0,0,0,0,2
325592,3355,274,44,162,65,3360,72,207,251,1727,...,0,0,0,0,0,0,0,1,0,1
326286,3355,268,43,134,49,3372,79,214,251,1697,...,0,0,0,0,0,0,0,1,0,1
352621,2836,268,43,300,202,1358,81,215,251,1482,...,0,0,0,0,0,0,0,0,0,2
353237,2804,264,44,270,170,1358,79,215,251,1443,...,0,0,0,0,0,0,0,0,0,2
353242,2913,265,42,420,279,1465,83,217,251,1584,...,0,1,0,0,0,0,0,0,0,2
355674,2741,264,44,210,103,1410,78,215,251,1348,...,0,0,0,0,0,0,0,0,0,2
357522,2778,279,47,240,137,1500,56,194,251,1354,...,0,0,0,0,0,0,0,0,0,2
358142,2711,276,43,180,73,1489,73,206,251,1289,...,0,0,0,0,0,0,0,0,0,2
356907,2859,267,42,330,225,1533,84,217,251,1448,...,0,0,0,0,0,0,0,0,0,2


Отсортируйте отфильтрованные данные: по убыванию

In [10]:
df_filtered.sort_values('Hillshade_3pm', ascending=False)

,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Hillshade_9am,Hillshade_Noon,Hillshade_3pm,Horizontal_Distance_To_Fire_Points,...,Soil_Type32,Soil_Type33,Soil_Type34,Soil_Type35,Soil_Type36,Soil_Type37,Soil_Type38,Soil_Type39,Soil_Type40,Cover_Type
392384,2925,275,50,120,32,2501,0,191,254,769,...,0,0,0,0,0,0,0,0,0,1
355675,2775,267,48,240,141,1429,60,204,254,1376,...,0,0,0,0,0,0,0,0,0,2
355676,2810,269,48,270,176,1448,56,201,254,1405,...,0,0,0,0,0,0,0,0,0,2
391753,2918,273,50,90,25,2496,0,192,254,780,...,0,0,0,0,0,0,0,0,0,1
354465,2801,272,47,270,167,1403,61,202,253,1423,...,0,0,0,0,0,0,0,0,0,2
356285,2769,266,47,240,131,1452,64,207,253,1368,...,0,0,0,0,0,0,0,0,0,2
355075,2804,276,48,270,170,1425,53,195,253,1414,...,0,0,0,0,0,0,0,0,0,2
357521,2741,273,48,210,97,1482,54,197,253,1324,...,0,0,0,0,0,0,0,0,0,2
358143,2744,276,50,210,100,1507,46,191,253,1318,...,0,0,0,0,0,0,0,0,0,2
358144,2785,274,47,240,141,1524,60,200,253,1347,...,0,0,0,0,0,0,0,0,0,2


## Анализ

Отфильтрованные данные отсортированы по возрастанию и убыванию признка, по которому происходила фильтрация.

Сортировка позволяет увидеть, что максимальные значения Hillshade_3pm (254) соответствуют участкам с наименьшей высотой солнца и, как правило, крутыми склонами и низкой утренней освещённостью (Hillshade_9am = 0).

Выполните группировку данных по одному категориальному признаку.

In [11]:
df_covtype = df.groupby('Cover_Type').mean()
df_covtype

,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Hillshade_9am,Hillshade_Noon,Hillshade_3pm,Horizontal_Distance_To_Fire_Points,...,Soil_Type31,Soil_Type32,Soil_Type33,Soil_Type34,Soil_Type35,Soil_Type36,Soil_Type37,Soil_Type38,Soil_Type39,Soil_Type40
Cover_Type,,,,,,,,,,,,,,,,,,,,,
1,3128.644888,156.138227,13.127110,270.555245,42.156939,2614.834517,211.998782,223.430211,143.875038,2009.253517,...,0.056000,0.100821,0.085668,0.000444,0.004395,0.000066,0.000000,0.041206,0.037207,0.022781
2,2920.936061,152.060515,13.550499,279.916442,45.884219,2429.530799,213.844423,225.326596,142.983466,2168.154849,...,0.046625,0.104327,0.089333,0.005051,0.000042,0.000148,0.000000,0.002612,0.001264,0.001172
3,2394.509845,176.372490,20.770208,210.276473,62.446915,943.940734,201.918415,215.826537,140.367176,910.955949,...,0.000000,0.002965,0.000140,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
4,2223.939934,137.139425,18.528941,106.934838,41.186749,914.199490,228.345832,216.997088,111.392792,859.124135,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
5,2787.417571,139.283051,16.641315,212.354893,50.610344,1349.765722,223.474876,219.035816,121.920889,1577.719794,...,0.032550,0.048457,0.054567,0.002107,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
6,2419.181897,180.539068,19.048886,159.853458,45.437439,1037.169805,192.844302,209.827662,148.284044,1055.351471,...,0.003628,0.011516,0.031036,0.000864,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
7,3361.928669,153.236226,14.255924,356.994686,69.474305,2738.250463,216.967723,221.746026,134.932033,2070.031594,...,0.010824,0.040907,0.031009,0.002487,0.046221,0.003072,0.014529,0.297611,0.271380,0.175134


    Рассчитайте для каждой группы:

    среднее значение числового признака;
    минимальное и максимальное значения;
    количество объектов в группе

In [12]:
df_group_one = df.groupby('Cover_Type')['Elevation'].agg(
    Size='size',
    Mean='mean',
    Minimum='min',
    Maximum='max'
)
df_group_one

,Size,Mean,Minimum,Maximum
Cover_Type,,,,
1,211840,3128.644888,2466,3686
2,283301,2920.936061,2142,3433
3,35754,2394.509845,1859,2899
4,2747,2223.939934,1988,2526
5,9493,2787.417571,2482,3011
6,17367,2419.181897,1863,2900
7,20510,3361.928669,2868,3858


## Анализ

Для выполнения группировки по одному категориальному признаку была выбрана целевая переменная Cover_Type. Она принимает 7 уникальных значений (1 - 7), что делает её удобным критерием для разбиения данных на группы и анализа различий между ними, в отличие от "прочих" категориальных "столбцов" (двух признаков, как было указано ранее) выбранного датасета. Все они являются бинарными и группировка по ним будет малоинформативна.

Дополнительно, была выполнена агрегация по признаку Elevation. Сделаны следующие выводы:

  - Самая многочисленная группа - класс 2 (283 301 объект).
  - Наибольшие значения показателя принимает класс 7 (3858).
  - Классы 1, 2 и 7 относятся к высокогорным (средние 2920–3362 м), класс 5 занимает промежуточное положение (2787 м), а классы 3, 4 и 6 — низкогорные (средние 2224–2419 м), при этом диапазоны высот внутри групп практически не пересекаются.

Выполните группировку по двум признакам.

In [13]:
df_two_feat = df.groupby(['Cover_Type', 'Wilderness_Area1']).mean() 
df_two_feat

Elevation      Aspect      Slope  \
Cover_Type Wilderness_Area1                                       
1          0                 3134.498780  175.049970  13.745305   
           1                 3122.768514  137.153854  12.506541   
2          0                 2935.928930  164.378822  15.573076   
           1                 2906.875702  140.508369  11.653721   
3          0                 2394.509845  176.372490  20.770208   
4          0                 2223.939934  137.139425  18.528941   
5          0                 2800.384804  164.875350  14.667192   
           1                 2767.827823  100.620471  19.623645   
6          0                 2419.181897  180.539068  19.048886   
7          0                 3378.988448  162.358038  14.401064   
           1                 3310.394825  125.681239  13.817487   

                             Horizontal_Distance_To_Hydrology  \
Cover_Type Wilderness_Area1                                     
1          0                                       302.483439   
           1                                       238.504432   
2          0                                       306.111689   
           1                                       255.350459   
3          0                                       210.276473   
4          0                                       106.934838   
5          0                                       236.890931   
           1                                       175.288019   
6          0                                       159.853458   
7          0                                       404.312155   
           1                                       214.059008   

                             Vertical_Distance_To_Hydrology  \
Cover_Type Wilderness_Area1                                   
1          0                                      50.788717   
           1                                      33.492012   
2          0                                      57.290641   
           1                                      35.187241   
3          0                                      62.446915   
4          0                                      41.186749   
5          0                                      52.461835   
           1                                      47.813277   
6          0                                      45.437439   
7          0                                      79.764618   
           1                                      38.389531   

                             Horizontal_Distance_To_Roadways  Hillshade_9am  \
Cover_Type Wilderness_Area1                                                   
1          0                                     2020.896695     207.370429   
           1                                     3211.053322     216.644910   
2          0                                     1673.239957     208.773048   
           1                                     3138.782697     218.600375   
3          0                                      943.940734     201.918415   
4          0                                      914.199490     228.345832   
5          0                                     1560.156688     216.759279   
           1                                     1031.925681     233.620206   
6          0                                     1037.169805     192.844302   
7          0                                     2658.555584     214.874554   
           1                                     2978.991178     223.290727   

                             Hillshade_Noon  Hillshade_3pm  \
Cover_Type Wilderness_Area1                                  
1          0                     224.193898     149.925784   
           1                     222.663592     137.801054   
2          0                     225.117918     147.097838   
           1                     225.522295     139.124996   
3          0                     215.826537     140.367176   
4          0                   

Рассчитайте агрегированные показатели:
какие комбинации признаков встречаются чаще всего;
какие группы можно считать аномальными или нетипичными.

In [14]:
df_group_two = df.groupby(['Cover_Type', 'Wilderness_Area1'])['Elevation'].agg(
    Size='size',
    Mean='mean',
    Minimum='min',
    Maximum='max'
)

df_group_two

Size         Mean  Minimum  Maximum
Cover_Type Wilderness_Area1                                       
1          0                 106123  3134.498780     2466     3489
           1                 105717  3122.768514     2518     3686
2          0                 137104  2935.928930     2142     3433
           1                 146197  2906.875702     2473     3333
3          0                  35754  2394.509845     1859     2899
4          0                   2747  2223.939934     1988     2526
5          0                   5712  2800.384804     2526     3011
           1                   3781  2767.827823     2482     2981
6          0                  17367  2419.181897     1863     2900
7          0                  15409  3378.988448     2868     3858
           1                   5101  3310.394825     3107     3619

## Анализ

##### Самые многочисленные

Самыми частыми можно считать комбинации (2,1), (2,0), (1,0), (1,1). Вместе они охватывают большую часть датасета, что подтверждает заявленнный дисбаланс.

##### Аномальные и нетипичные

1. Из 14 теоретически возможных комбинаций (7 классов × 2 значения Wilderness_Area1) в датасете присутствуют 11. Отсутствуют три комбинации: (3, 1), (4, 1), (6, 1), то есть классы 3, 4 и 6 не встречаются в первой зоне дикой природы.
2. Малые группы (5, 1), (4, 0) статистически нетипичны из-за малого числа объектов.
3. Комбинация (4, 0) имеет сравнительно низкие значения по всем показателям.

Используя данные из лабораторной:

объедините две таблицы с помощью merge или объедините несколько таблиц с помощью concat.

    

In [15]:
data1 = df.head(50)
cover_type_info = pd.DataFrame({
    'Cover_Type': [1, 2, 3, 4, 5, 6, 7],
    'Avg_Temperature_C': [4.5, 5.2, 8.1, 11.3, 7.8, 9.4, 2.1],
})
df_merged = pd.merge(data1, cover_type_info, on='Cover_Type', how='left')
df_merged

,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Hillshade_9am,Hillshade_Noon,Hillshade_3pm,Horizontal_Distance_To_Fire_Points,...,Soil_Type33,Soil_Type34,Soil_Type35,Soil_Type36,Soil_Type37,Soil_Type38,Soil_Type39,Soil_Type40,Cover_Type,Avg_Temperature_C
0,2596,51,3,258,0,510,221,232,148,6279,...,0,0,0,0,0,0,0,0,5,7.8
1,2590,56,2,212,-6,390,220,235,151,6225,...,0,0,0,0,0,0,0,0,5,7.8
2,2804,139,9,268,65,3180,234,238,135,6121,...,0,0,0,0,0,0,0,0,2,5.2
3,2785,155,18,242,118,3090,238,238,122,6211,...,0,0,0,0,0,0,0,0,2,5.2
4,2595,45,2,153,-1,391,220,234,150,6172,...,0,0,0,0,0,0,0,0,5,7.8
5,2579,132,6,300,-15,67,230,237,140,6031,...,0,0,0,0,0,0,0,0,2,5.2
6,2606,45,7,270,5,633,222,225,138,6256,...,0,0,0,0,0,0,0,0,5,7.8
7,2605,49,4,234,7,573,222,230,144,6228,...,0,0,0,0,0,0,0,0,5,7.8
8,2617,45,9,240,56,666,223,221,133,6244,...,0,0,0,0,0,0,0,0,5,7.8
9,2612,59,10,247,11,636,228,219,124,6230,...,0,0,0,0,0,0,0,0,5,7.8


In [16]:
print(f"Shape: {df_merged.shape}")
print(df_merged.isna().sum().sum())

Shape: (50, 56)
0


## Анализ

Размер итоговой таблицы составил 50 строк и 56 столбцов.

Объединение выполнялось по ключу Cover_Type - это общий столбец для обеих таблиц, а how='left' задал левое объединение: используются ключи из левого датасета.

Из справочной таблицы cover_type_info добавлен один новый столбец — Avg_Temperature_C со средней температурой произрастания для каждого типа лесного покрова. Эти данные присоединились к каждой из 50 строк data1 в соответствии со значением Cover_Type в этой строке.

Пропущенных значений не обнаружено. Это объясняется тем, что:

  - Ключ Cover_Type присутствует в обеих таблицах
  - Все значения ключа из data1 есть в справочнике

## Итоги

В лабораторной работе использовался датасет Forest Cover Types с площадки Kaggle, содержащий информацию о типах лесного покрова для территорий национального парка Рузвельта в Колорадо. Датасет включает 581 012 записей и 55 столбцов, среди которых 10 непрерывных числовых признаков, 44 бинарных категориальных признака и одна целевая переменная Cover_Type с 7 классами. Все столбцы имеют тип int64, пропущенных значений нет, что делает датасет хорошо подготовленным для анализа. Особенностью структуры является то, что стандартное разделение на числовые и категориальные признаки по типу данных здесь не работает, поэтому категориальные признаки были выделены эвристическипо количеству уникальных значений. Числовые признаки описывают высоту, уклон, ориентацию склона, расстояния до воды, дорог и пожаров, а также освещённость в разное время суток. Категориальные признаки представлены бинарными индикаторами зон дикой природы и типов почв. Целевая переменная Cover_Type принимает 7 значений и характеризуется сильным дисбалансом классов. Наиболее полезными операциями pandas в ходе работы оказались: read_csv для загрузки данных, head, shape и dtypes для первичного обзора, nunique для разделения признаков, describe, mean и std для описательной статистики, а также фильтрация и sort_values для сортировки. Отдельно стоит отметить метод groupby с агрегацией через agg, который позволил рассчитать статистику по группам как по одному признаку, так и по двум. Метод merge оказался незаменим для присоединения справочной информации по общему ключу Cover_Type. В целом, pandas предоставил полный набор инструментов для анализа табличных данных - от загрузки и очистки до группировки и объединения.

## Выводы:

* Данные можно характеризовать как дисбалансные. Классы 1 и 2 доминируют над числом других, а числовые признаки имеют разный масштаб.
* Высота является сильным разделяющим признаком между классами. Также группы существенно различаются по освещённости, расстояниям до инфраструктуры и почвенному составу. Классы 1, 2 и 7 можно назвать высокогорными, классы 3, 4 и 6 - к низкогорными, а класс 5 занимает промежуточное положение. Классы 1 и 2 наиболее удалены от дорог, а классы 3 и 4 — наименее. Класс 7 характеризуется высокой долей почв типов 38, 39 и 40.
* Данные являются полными, так как пропущенных значений нет. Структура требует внимания, так как данные не различимы по типу, а сильный дисбаланс может отрицательно сказаться на качестве обучения, особенно с учетом того, что здесь присутствуют малочисленные классы. Таким образом, датасет требует некоторой доработки, как минимум масштабирования числовых признаков.